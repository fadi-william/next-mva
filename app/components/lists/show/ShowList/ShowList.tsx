// The imported libs.
import { observer } from "mobx-react";
import * as React from "react";

// The component's model type.
import { TShowList } from "../../../../models/show/ShowList";

// The external imported components.
import ShowListItem from "../ShowListItem/ShowListItem";

// The ShowList's props interface.
interface IShowListProps {
    showList: TShowList;
}

@observer
export default class ShowList extends React.Component<IShowListProps> {

    public render() {
        const { shows, isLoading } = this.props.showList;

        return (
            <div>
                <h1>Shows</h1>
                {!isLoading && <ul>
                    {shows.map((show, idx) => (
                        <ShowListItem key={idx} show={show} />
                    ))}
                </ul>}
                {isLoading && <div>
                    Loading...
                </div>}
            </div>
        );
    }
}
