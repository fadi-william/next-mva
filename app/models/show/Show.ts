// The imported libs.
import { applySnapshot, flow, types } from "mobx-state-tree";
import { getShow } from "../../service/showService";

export type TShow = typeof Show.Type;

// The module's global variables.
let store: TShow = null;

// If the id changed, we will re-initialize the store *(on the client-side).
let currentShowId: number = null;

export const Show = types.model({
    id: types.maybe(types.identifier(types.number)),
    image: types.maybe(types.string),
    isLoading: types.optional(types.maybe(types.boolean), true),
    name: types.maybe(types.string),
    summary: types.maybe(types.string),
})
.actions( (self) => ({
    doneLoading() {
        self.isLoading = false;
    },
}))
.actions( (self) => ({
    getShow: flow(function *() {
        const response = yield getShow(self.id);

        if (response.data) {
            const show = Object.assign({}, response.data);

            show.id = Number(response.data.id);
            show.image = response.data.image.medium;

            applySnapshot(self, show);

            self.doneLoading();
        }
    }),
}))
.actions( (self) => ({
    afterCreate() {
        self.getShow();
    },
}));

// Utility function - Instantiates a part of the mobx-state-tree store.
export function getStore(isServer, showId, snapshot = null) {
    // If this is the server.
    if (isServer) {
      currentShowId = showId;
      store = Show.create({
          id: showId,
      });
    }
    if (store === null || currentShowId !== showId) {
        currentShowId = showId;
        store = Show.create({
            id: showId,
        });
    }
    if (snapshot) {
      applySnapshot(store, snapshot);
    }
    return store;
}
