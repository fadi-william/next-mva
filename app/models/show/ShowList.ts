// The imported libs.
import { applySnapshot, types  } from "mobx-state-tree";
import "rxjs/add/operator/debounceTime";
import "rxjs/add/operator/map";
import { Subject } from "rxjs/Subject";

import { getShows } from "../../service/showService";
import { Show } from "./Show";

export type TShowList = typeof ShowList.Type;

// The module's global variables.
let store: TShowList = null;
const getShowsSubject$ = new Subject<void>();

export const ShowList = types.model({
    isLoading: types.optional(types.maybe(types.boolean), true),
    shows: types.optional(types.array(Show), []),
})
// The shows model's related actions.
.actions( (self) => ({
    doneLoading() {
        self.isLoading = false;
    },
    getShows() {
        // Get the latest shows.
        getShowsSubject$.next();
    },
}))
.actions( (self) => ({
    afterCreate() {
        getShowsSubject$
            .debounceTime(200)
            .map(() => {
                return getShows();
            })
            .subscribe((showsService) => {
                showsService.then((showResponse) => {
                    const shows = showResponse.data.map( (showObj) => {
                        return {
                            id: showObj.show.id,
                            name: showObj.show.name,
                        };
                    });

                    applySnapshot(self.shows, shows);

                    if (self.isLoading) {
                        self.doneLoading();
                    }
                });
            });

        // Trigger get shows.
        self.getShows();
    },
    beforeDestroy() {
        getShowsSubject$.unsubscribe();
    },
}))
// Derived data from the shows model.
.views( (self) => ({
    get totalShows() {
        return self.shows.length;
    },
}));

// Utility function - Instantiates a part of the mobx-state-tree store.
export function getStore(isServer, snapshot = null) {
    // If this is the server.
    if (isServer) {
      store = ShowList.create();
    }
    if (store === null) {
        store = ShowList.create();
    }
    if (snapshot) {
      applySnapshot(store, snapshot);
    }
    return store;
}
