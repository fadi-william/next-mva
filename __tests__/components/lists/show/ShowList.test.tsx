import * as React from "react";
import * as renderer from "react-test-renderer";
import { getStore } from "../../../../app/models/show/ShowList";

import ShowList from "../../../../app/components/lists/show/ShowList/ShowList";

// Mock the show service.
jest.mock("../../../../app/service/showService");

it("show list item component while loading - snapshot test", () => {
    const showStore = getStore(false);
    const tree = renderer.create(
        <ShowList showList={showStore} />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
});

it("show list item component after loading - snapshot test", async () => {
    const showStore = getStore(false);
    await showStore.afterCreate();
    await showStore.getShows();

    let tree;
    await new Promise( (res) => setTimeout(() => {
        tree = renderer.create(
            <ShowList showList={showStore} />,
        ).toJSON();
        res();
    }, 3000));

    expect(tree).toMatchSnapshot();
});
