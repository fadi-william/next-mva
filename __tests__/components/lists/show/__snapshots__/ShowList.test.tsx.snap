// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show list item component after loading - snapshot test 1`] = `
<div>
  <h1>
    Shows
  </h1>
  <ul>
    <li>
      <a
        href="/p/975"
        onClick={[Function]}
      >
        Batman
      </a>
    </li>
    <li>
      <a
        href="/p/481"
        onClick={[Function]}
      >
        The Batman
      </a>
    </li>
    <li>
      <a
        href="/p/504"
        onClick={[Function]}
      >
        Batman Beyond
      </a>
    </li>
  </ul>
</div>
`;

exports[`show list item component while loading - snapshot test 1`] = `
<div>
  <h1>
    Shows
  </h1>
  <div>
    Loading...
  </div>
</div>
`;
